roxie/ccd/ccdactivities.cpp

```diff
@@ -2661,29 +2662,61 @@ class CRoxieIndexActivity : public CRoxieKeyedActivity
     bool sendContinuation(IMessagePacker * output)
     {
         MemoryBuffer si;
-        unsigned short siLen = 0;
+        unsigned siLen = 0;
         si.append(siLen);
-        si.append(lastRowCompleteMatch);
+        si.append(false);
+        si.append(lastRowCompleteMatch);  // Note - this flag is also read on server side in isCompleteMatchFlag so (a) be careful about moving it and (b) don't compress it
         si.append(inputsDone);
         si.append(processed);
         si.append(keyprocessed);
         si.append(lastPartNo.partNo);
         si.append(lastPartNo.fileNo);
         tlk->serializeCursorPos(si);
-        if (si.length() <= maxContinuationSize)
+        if (si.length() <= continuationCompressThreshold || (continuationCompressThreshold==0 && si.length() <= continuationErrorThreshold))
         {
             siLen = si.length() - sizeof(siLen);
             si.writeDirect(0, sizeof(siLen), &siLen);
             output->sendMetaInfo(si.toByteArray(), si.length());
             return true;
         }
+        else if (continuationCompressThreshold)
+        {
+            MemoryBuffer compressed;
+            compressed.append(siLen);  // Leaving space to patch when size known
+            compressed.append(true);
+            compressed.append(lastRowCompleteMatch);  // This field is not compressed - see above!
+            compressToBuffer(compressed, si.length() - compressed.length(), si.toByteArray() + compressed.length());
+            bool report = logctx.queryTraceLevel() && (logctx.queryTraceLevel() > 3 || si.length() >= continuationWarnThreshold);
+            if (report)
+                DBGLOG("ERROR: continuation data size %u for %d cursor positions is large - compressed to %u", si.length(), tlk->numActiveKeys(), compressed.length());
```

Comment:
I chose ERROR deliberately - I consider this case to represent ECL code that is so dangerously inefficient that Roxie would have been smart to treat it as an error from the start. Unfortunately since there are existing usages out there that rely on it, I can't make it fatal.